### ISSUE TITLE
@nrwl/angular:component error in v14.6.* and v14.7.*

### ISSUE THREAD

**Issue by ElecTreeFrying** [nrwl/nx#11919]
## Current Behavior

I can't generate component using `@nrwl/angular:component`.

## Expected Behavior

Generate angular component successfully.

## Steps to Reproduce

1. `npx create-nx-workspace test`
2. choose `angular-nest`
3. run `nx g nrwl/angular:library reproduce` / `nx g lib reproduce`
4. In `libs/reproduce/src/lib` run `nx g @nrwl/angular:component reproduce` / `nx g component reproduce`

### Failure Logs

`>  NX   The path provided for the component (libs/reproduce/src/lib) does not exist under the project root (apps/my-app). Please make sure to provide a path that exists under the project root.`

### Environment

```
   Node : 14.18.1
   OS   : win32 x64
   npm  : 6.14.15

   nx : 14.6.5
   @nrwl/angular : 14.6.5
   @nrwl/cypress : 14.6.5
   @nrwl/detox : Not Found
   @nrwl/devkit : 14.6.5
   @nrwl/eslint-plugin-nx : 14.6.5
   @nrwl/express : Not Found
   @nrwl/jest : 14.6.5
   @nrwl/js : 14.6.5
   @nrwl/linter : 14.6.5
   @nrwl/nest : 14.6.5
   @nrwl/next : Not Found
   @nrwl/node : 14.6.5
   @nrwl/nx-cloud : 14.6.1
   @nrwl/nx-plugin : Not Found
   @nrwl/react : Not Found
   @nrwl/react-native : Not Found
   @nrwl/schematics : Not Found
   @nrwl/storybook : 14.6.5
   @nrwl/web : 14.6.5
   @nrwl/workspace : 14.6.5
   typescript : 4.7.3
   ---------------------------------------
   Local workspace plugins:
   ---------------------------------------
   Community plugins:
         @nguniversal/express-engine: 14.1.0
         @nguniversal/builders: 14.1.0
```

**Comment by junekpavel** [comment]
Hi, I have exactly same issue using nx `14.6.4`.

**Comment by leosvelperez** [comment]
Thanks for reporting this!

Unfortunately, I can reproduce it. Could you provide more detailed steps by steps info to reproduce the issue? Exact command used, whether it happens when running from the terminal directly or using Nx Console, etc.

**Comment by junekpavel** [comment]
Little more information, I just figured out.

The generate command apparently needs to specify project where the component should be generated. So for instance, the following command will generate component properly in the desired library:
```
nx g c components/button --project=ui-kit --export
```

When you omit the project argument, the component is generated into the default project:
```
nx g c components/button --export
```

---

These both examples described above are behaving as expected. 

But there is one difference, in comparison to previous versions of nx. When you don't specify the `project` argument, but rather `path` one. The example command is following:

```
nx g c button --path=libs/ui/kit/src/lib/components --export
```

In previous version, this command would generate new component in the library. The result would be equal to running first described command. But now, instead of generating new component, I see error:

```
 The path provided for the component (libs/ui/kit/src/lib/components) does not exist under the project root (apps/my-app). Please make sure to provide a path that exists under the project root.
```

---

To add more context there, why the `path` argument is used instead of `project`.

The WebStorm IDE have pretty good Angular support, which includes action to generate new component in desired directory by just few clicks. You right-click to directory, select you want to generate new component, and type it's name. That action run command which uses the `path` argument.

![CleanShot 2022-09-09 at 12 57 43@2x](https://user-images.githubusercontent.com/39130083/189336123-4a7cd644-4cc2-488a-a806-347803ce33a2.png)

**Comment by leosvelperez** [comment]
I see. Thanks for the details provided!

This is by design. There's a check on the component generator to make sure the path provided belongs to the project where it's supposed to go. This is to avoid mistakes with wrong paths, it's a safety mechanism to ensure no random path is provided. Without that check, you could provide a misspelled path like `apps/app2/src/app` for the `app1` project. When a project is not provided the default project is used, so the check runs against it.

With Nx Console, the project is auto-populated when running such a generator, so you normally wouldn't have that issue. It might be a good feature request for WebStorm plugin and in the meantime, you just need to make sure to specify the right project you want to generate your component in.

I'm going to close this since it's working as designed.

**Comment by ElecTreeFrying** [comment]
Please reopen this issue. @leosvelperez 

I reverted back to v14.5.10, `nx g component` works fine and normally. I think something is wrong with v14.6.x releases with `@nrwl/angular:component`.

**Comment by junekpavel** [comment]
Thank for your response.

Could you please tell us, at which version was this "feature" introduced (didn't find it in the changelog), so we could downgrade to that version and keep using it until this issue is somehow resolved?

Because I don't want to make sure I specify right project. I want to select directory, type component name and want my component to be generated (this takes about few seconds). In comparison to need of type-in project name (by memory), type directory in that project (also by memory, or I can copy the path to repo root and delete half of it), and finally type component name. This process takes much more time than previous workflow.

---

But I really see this as a step-back and if this is working as designed, could you please explain us the design?

As I understand it, nx is presenting itself as a replacement for Angular CLI. During the installation of nx, the Angular CLI is disabled in favor of nx and the `ng` command is aliased to `nx`. So I would expect that every `nx` command which is replacement to `ng` would work the very same way as the `ng` one.

So I would expect exactly same result, if I run this command in plain Angular CLI project and in nx project:

```
ng g c button --path=libs/ui/kit/src/lib/components
```

But the result is not same, as nx throw an error. So I'd suppose, this is bug in nx. And if it's not, I do believe that the `ng` command should't be aliased at all. It should be up to developer which command want to to use.

---

Also the `nx generate component` is so far the only command which is broken. Other Angular generators works fine, for example, module or directive generators works as desired. If I run the following command, the directive is generated without any problems. 

```
nx g directive button-addon --path=libs/ui/kit/src/lib/components/button
```

**Comment by ElecTreeFrying** [comment]
Steps to reproduce
1. `npx create-nx-workspace test`
2. choose `angular-nest`
3. `nx g nrwl/angular:library reproduce`
4. In `libs/reproduce/src/lib` run `nx g @nrwl/angular:component reproduce`

This is the error
`>  NX   The path provided for the component (libs/reproduce/src/lib) does not exist under the project root (apps/my-app). Please make sure to provide a path that exists under the project root.`

> I see. Thanks for the details provided!
> 
> This is by design. There's a check on the component generator to make sure the path provided belongs to the project where it's supposed to go. This is to avoid mistakes with wrong paths, it's a safety mechanism to ensure no random path is provided. Without that check, you could provide a misspelled path like `apps/app2/src/app` for the `app1` project. When a project is not provided the default project is used, so the check runs against it.
> 
> With Nx Console, the project is auto-populated when running such a generator, so you normally wouldn't have that issue. It might be a good feature request for WebStorm plugin and in the meantime, you just need to make sure to specify the right project you want to generate your component in.
> 
> I'm going to close this since it's working as designed.

**Comment by ElecTreeFrying** [comment]
Upon testing the error is also showing in `v14.7.0` @leosvelperez

**Comment by ElecTreeFrying** [comment]
Please reopen this issue. @leosvelperez

**Comment by junekpavel** [comment]
Hey, @ElecTreeFrying.

I temporarily fixed this by manually patching nx generator via [patch-package](https://www.npmjs.com/package/patch-package) to disable `path` argument validation.

Now the generator is working as expected for me. You can give it a try, if you haven't find better solution already.

The patch is following:

```patch
diff --git a/node_modules/@nrwl/angular/src/generators/component/component.js b/node_modules/@nrwl/angular/src/generators/component/component.js
index 88809ff..dd13d82 100644
--- a/node_modules/@nrwl/angular/src/generators/component/component.js
+++ b/node_modules/@nrwl/angular/src/generators/component/component.js
@@ -11,7 +11,8 @@ function componentGenerator(tree, rawOptions) {
     return tslib_1.__awaiter(this, void 0, void 0, function* () {
         const options = (0, normalize_options_1.normalizeOptions)(tree, rawOptions);
         const { projectSourceRoot } = options, schematicOptions = tslib_1.__rest(options, ["projectSourceRoot"]);
-        checkPathUnderProjectRoot(tree, options);
+        // Disable path checking so we can generate components in any project, see https://github.com/nrwl/nx/issues/11919
+        // checkPathUnderProjectRoot(tree, options);
         const angularComponentSchematic = (0, ngcli_adapter_1.wrapAngularDevkitSchematic)('@schematics/angular', 'component');
         yield angularComponentSchematic(tree, schematicOptions);
         (0, component_1.exportComponentInEntryPoint)(tree, options);

```

**Comment by leosvelperez** [comment]
@ElecTreeFrying, @junekpavel thanks for providing more details about the pain points.

As explained before this is a check that we do to ensure the path being passed belongs to the specified project, to avoid issues related to those properties not being in sync. We use the project and make decisions based on the configuration of the project, so, if the project and the path are not in sync some issues might arise.

That said, I can see how restrictive it might be in some scenarios. If someone just wants to specify only the path to the component, we should support it. We do need the project, as I mentioned above, but we could infer the project based on the provided path only if the project was not provided. We would keep the check if both properties are provided. So, the check is still needed but only in certain scenarios.

Once again, thanks for raising this and sticking to it! We'll work on addressing it.

P.S. Just a quick note to clarify some points made above:

> As I understand it, nx is presenting itself as a replacement for Angular CLI. During the installation of nx, the Angular CLI is disabled in favor of nx and the `ng` command is aliased to `nx`. So I would expect that every `nx` command which is replacement to `ng` would work the very same way as the `ng` one.

Nx is not just a "replacement" for the Angular CLI. It's that and much more. The fact that Nx can replace the Angular CLI (it does offer most of the commands of the Angular CLI) doesn't mean that it has to do everything the same. The point is to offer a superior CLI experience.

There also seems to be confusion between the boundaries of the CLI and the generators offered by plugins. The `generate` command behaves pretty similar between both Nx and Angular CLI, though they have different implementations. What's at fault here is not the command nor the CLI used, it's the generator. The `@nrwl/angular:component` generator checks that the path is contained in the specified project. The `@schematics/angular:component` doesn't.

> But the result is not same, as nx throw an error. So I'd suppose, this is bug in nx. And if it's not, I do believe that the `ng` command should't be aliased at all. It should be up to developer which command want to to use.

As explained above, the issue is not the command (`generate`) nor the CLI (`nx` vs `ng`), but the generator/schematic used. You can still run `nx g @schematics/angular:component` instead of the `@nrwl/angular:component`. You don't need to change the CLI that you use for that, Nx can run Angular CLI schematics. In fact, when you run the `@nrwl/angular:component` generator, the `@schematics/angular:component` is run under the hood for the base generation, and then we add some features on top of it.

As also explained, the check for the path was a decision made to prevent misusing the generator. Right now we need it for all scenarios because we rely on the default project to populate the project property when none is specified. Based on the discussion here, we could get rid of it when the project option is not provided by inferring it from the path instead of getting the default project.

**Comment by ElecTreeFrying** [comment]
More context about my nx workspace

My default collection is set to `@nrwl/angular`, this is why I use `nx g component [component-name]` syntax
```
  "cli": {
    "defaultCollection": "@nrwl/angular",
    "packageManager": "npm"
  },
```

I started this project with nx `v12.6.6`, and ever since I'm always updating my monorepo to the latest versions using `nx migrate latest`. As mentioned above, in `v14.5.*` (and versions before) everything's working perfectly fine, not until `v14.6.*^`

**Comment by Coly010** [comment]
Duplicate of https://github.com/nrwl/nx/issues/11883

It is pretty straightforward to support this where you can just be in the same folder in the terminal as the project you want to generate the component in. 

But ideally, the `--project` flag should be provided

**Comment by ElecTreeFrying** [comment]
If you have a `defaultProject` in your nx.json all commands points to it. You use `--project` flag if you want to target another application

In my case, the `defaultProject` is set to my main front-end app, that's why I don't need `--project` flag, unless I want to create a lib for another application.

**Comment by Coly010** [comment]
Or library. Libraries still count as different projects.

If you have an 3 apps and 3 libraries, and you want to generate a a component in one of the libraries, but your defaultProject is pointing to an app, then you should provide `--project` for the library you want to create the component in.

**Comment by leosvelperez** [comment]
Just to clarify, when https://github.com/nrwl/nx/pull/11980 lands, you won't need to provide `--project` if only the `--path` is provided in the right place. The `project` will be inferred from the `path` in those cases. If you do provide `--project` and `--path`, we'll check whether the `path` is contained inside the `project` root.
